### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by BenXu86**
SWE model
Support megapixel level emulation using MAC grid.

### CODE REVIEW

File: Source/Dynamics/HeightField/ShallowWaterEquationModel.cu [hunk 1]
```diff
@@ -57,221 +69,399 @@ namespace PhysIKA
...
+					break;
+				case 2:
+					m_velocity[i][0] = 0;
+					h[i] = h[(ix + 1)*zcount + iz];
+					break;
+				case 3:
+					m_velocity[i][0] = 0;
+					h[i] = h[(ix - 1)*zcount + iz];
+					break;
+				}
```
Comment (FeiZhu) [inline Source/Dynamics/HeightField/ShallowWaterEquationModel.cu:147]: for loop, if statement, switch statement are highly unfriendly to CUDA. Try to remove if and switch by some tricky manipulation. For instance,
int ix_offset = j < 2 ? 0 : ( j % 2 ? -1 : 1);
int iz_offset = j < 2 ? (j == 1 ? -1 : 1) : 0;
m_velocity[i][2] = 0;
h[i] = h[(ix + ix_offset) * zcount + iz + iz_offset];

**Review by FeiZhu — CHANGES_REQUESTED** [review]

File: Source/IO/Image_IO/image.cpp [hunk 2]
```diff
@@ -115,6 +115,31 @@ void Image::flipVertically()
...
 }
 
+void Image::mergeImage(int h_compressed, int w_compressed)
+{
+	int pixel_size = pixelSize();
+	unsigned int height = height_ / h_compressed;//a new pixel is calculate by h_compressed*w_compressed pixels
+	unsigned int width = width_ / w_compressed;
+	
+	unsigned char *d = new unsigned char[width*height*pixel_size];
+	for (unsigned int i = 0; i < height; ++i)
```
[the remark this answers is not in the document]
Comment (BenXu86) [reply Source/IO/Image_IO/image.cpp:125]: We only use this function once to do image preprocessing,  and we won’t use it again.

File: Source/Rendering/HeightFieldRender.cu [hunk 3]
```diff
@@ -95,32 +95,34 @@ namespace PhysIKA
...
+		float3 colorWater,
+		float3 colorTerrain)
 	{
 		int i = threadIdx.x + blockIdx.x * blockDim.x;
 		int j = threadIdx.y + blockIdx.y * blockDim.y;
 
 		if (i < heights.Nx() - 1 && j < heights.Ny() - 1)
 		{
-			int id = i + j * heights.Nx();
+			int id = i + j * (heights.Nx() - 1);
```
[the remark this answers is not in the document]
Comment (BenXu86) [reply Source/Rendering/HeightFieldRender.cu:110]: yes

File: Source/Rendering/HeightFieldRender.cu [hunk 4]
```diff
@@ -175,28 +177,41 @@ namespace PhysIKA
...
 			normals,
 			colors,
 			heights,
+			terrain,
 			hf->getDx(),
 			hf->getDz(),
 			make_float3(ori[0], ori[1], ori[2]),
+			make_float3(0.0, 0.2, 1),
 			make_float3(1.0, 0.0, 0.0));
+		char str[200];
```
[the remark this answers is not in the document]
Comment (BenXu86) [reply Source/Rendering/HeightFieldRender.cu:205]: The synchronization seems unnecessary. But the rendering part is not written by us, so I'm not sure about this. I suggest to keep it like this for now.

File: Source/Framework/Framework/DeclareEnum.h [hunk 5]
```diff
@@ -1,6 +1,6 @@
 #pragma once
 #include <map>
-
+#include <string>
```
[the remark this answers is not in the document]
Comment (BenXu86) [reply Source/Framework/Framework/DeclareEnum.h:3]: This file was not changed by us. But the ‘string’ variable in this header file is used by the following class, so it is necessary to include it.

File: Source/Dynamics/HeightField/ShallowWaterEquationModel.cu (hunk 1 above)
Comment (BenXu86) [reply Source/Dynamics/HeightField/ShallowWaterEquationModel.cu:147]: Thanks a lot. I learn it. I will make some optimizations for this in particular

File: Source/IO/Image_IO/image.h [hunk 6]
```diff
@@ -45,11 +45,12 @@ class Image
     //image operations
     void flipHorizontally();
     void flipVertically();
+    void mergeImage(int h_compressed, int w_compressed);
```
[the remark this answers is not in the document]
Comment (BenXu86) [reply Source/IO/Image_IO/image.h]: Now we add the explaination in the function declaration

File: Source/Dynamics/HeightField/HeightFieldNode.cu [hunk 7]
```diff
@@ -10,9 +12,10 @@ namespace PhysIKA
 	HeightFieldNode<TDataType>::HeightFieldNode(std::string name = "default")
 		: Node(name)
 	{
-		auto swe = this->template setNumericalModel<ShallowWaterEquationModel<TDataType>>("swe");
-		this->setNumericalModel(swe);
-		SWEconnect();
+		auto swe = this->setNumericalModel<ShallowWaterEquationModel<TDataType>>("swe");
+		//������std::shared_ptr<Module>
+		//this->setNumericalModel(swe);
```
[the remark this answers is not in the document]
Comment (BenXu86) [reply Source/Dynamics/HeightField/HeightFieldNode.cu]: we have deleted the unnecessary comments now.

File: Source/Dynamics/HeightField/HeightFieldNode.cu [hunk 8]
```diff
@@ -317,7 +293,7 @@ namespace PhysIKA
 	}
 
 	template<typename TDataType>
-	std::vector<TDataType::Real> HeightFieldNode<TDataType>::outputDepth() {
+	std::vector<TDataType::Real>& HeightFieldNode<TDataType>::outputDepth() {
```
Comment (FeiZhu) [inline Source/Dynamics/HeightField/HeightFieldNode.cu:296]: It is danger to return reference of local variable.

File: Source/IO/Image_IO/image.cpp [hunk 9]
```diff
@@ -136,8 +137,9 @@ void Image::mergeImage(int h_compressed, int w_compressed)
 			}
 	height_ = height;
 	width_ = width;
-	allocMemory();
-	memcpy(raw_data_ ,d, sizeof(unsigned char)*pixel_size*height_*width_);
+    if (raw_data_)
+        delete[] raw_data_;
+    raw_data_ = d.data();
```
Comment (FeiZhu) [inline Source/IO/Image_IO/image.cpp:142]: d cannot be std::vector considering its memory will be pointed by raw_data_. It must be allocated on heap.

**Review by FeiZhu — CHANGES_REQUESTED** [review]

**Pull request merged**